--- socketft.h.orig	2011-05-17 12:38:55.671875000 +0200
+++ socketft.h	2011-05-17 12:39:01.796875000 +0200
@@ -92,8 +92,8 @@
 	void CheckAndHandleError(const char *operation, socket_t result) const
 		{if (result == SOCKET_ERROR) HandleError(operation);}
 #ifdef USE_WINDOWS_STYLE_SOCKETS
-	void CheckAndHandleError(const char *operation, BOOL result) const
-		{assert(result==TRUE || result==FALSE); if (!result) HandleError(operation);}
+	//void CheckAndHandleError(const char *operation, BOOL result) const
+	//	{assert(result==TRUE || result==FALSE); if (!result) HandleError(operation);}
 	void CheckAndHandleError(const char *operation, bool result) const
 		{if (!result) HandleError(operation);}
 #endif
